flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManager.java

```diff
@@ -302,7 +302,12 @@ public boolean unregisterSlotRequest(AllocationID allocationId) {
 		PendingSlotRequest pendingSlotRequest = pendingSlotRequests.remove(allocationId);
 
 		if (null != pendingSlotRequest) {
-			cancelPendingSlotRequest(pendingSlotRequest);
+			if (pendingSlotRequest.isAssigned()) {
+				cancelPendingSlotRequest(pendingSlotRequest);
+			}
+			else {
+				resourceActions.cancelResourceAllocation(pendingSlotRequest.getResourceProfile());
```

Comment:
I think we should not immediately cancel ongoing resource allocations. The `SlotManager` could decide upon registration of a new worker whether this one is actually needed or not. In the latter case it could release the resource. This would also simplify the protocol since you don't know whether you still can cancel an ongoing resource allocation.

Reply:
Yes, the SlotManager can decide to release the resource more than needed. But in a worst case:
1. Now the MESOS or YARN cluster have not enough resource.
2. A job ask for 100 worker of size A;
3. As there are not enough resource, the job failover, the previous 100 is not cancelled, it ask another 100.
4. This repeated several times, the pending requests for worker of size A reaches 10000.
5. A worker of size B crashed, so the job now only need 100 woker of size A and 1 worker of size B. But the YARN or MESOS think the job need 10000 A and 1 B as the request are never cancelled.
6. The MESOS/YARN now have some resources for 110 A, more than 100 A and 1 B, and it begin to assign resource for the job, but it first try to allocate 10000 containers of size A, and the job still can not be started as it is lack of container B. 
7. This may cause the job can not be started even when the cluster resource is now enough in a long time.
8. And this did happen in our cluster, as our cluster is busy. So I think it's better to keep this protocol, and different resource managers can treat this protocol according to their need.

Reply:
Then this should be added as a separate feature because it is not strictly required by this PR here.

Furthermore, I'm not sure whether this shouldn't be the responsibility of the `ResourceManager`. E.g. we could think about adding a timeout for container requests after which we cancel them. Additionally, if we add support for starting machines with multiple slots, then we shouldn't release a requested resource if only a single of it slots is no longer needed. That is something else to consider before adding the cancel resource allocation method.